housekeeping/src/main/groovy/whelk/housekeeping/NotificationSender.groovy

```diff
@@ -0,0 +1,428 @@
+
+        List<String> triggeredTriggers = []
+
+        user.requestedNotifications.each { request ->
+
+            // This stuff (the request) comes from a user, so we must be super paranoid about it being correctly formed.
+
+            if (! request instanceof Map)
+                return
+
+            if (! request["heldBy"] instanceof String)
+                return
+
+            if (request["heldBy"] != heldBy)
+                return
+
+            if (! request["triggers"] instanceof List)
+                return
+
+            for (Object triggerObject : request["triggers"]) {
+                if (! triggerObject instanceof String)
+                    return
+                String triggerUri = (String) triggerObject
+                if (triggerIsTriggered(instanceId, from, until, triggerUri))
+                    triggeredTriggers.add(triggerUri)
+            }
+        }
+
+        return triggeredTriggers
+    }
+
+    private boolean triggerIsTriggered(String instanceId, Instant before, Instant after, String triggerUri) {
+        Document instanceAfterChange = whelk.getStorage().loadAsOf(instanceId, Timestamp.from(after))
+
+        // Populate 'changeNotes' with every note affecting this instance in the specified interval
+        List<Map> changeNotes = new ArrayList<>();
+        List notesOnInstance = (List) Document._get(["@graph", 0, "hasChangeNote"], instanceAfterChange.data)
+        if (notesOnInstance != null)
+        changeNotes.addAll(notesOnInstance)
+        switch (triggerUri) {
+            case "https://id.kb.se/changenote/primarytitle":
+            case "https://id.kb.se/changenote/maintitle":
+                historicEmbellish(instanceAfterChange, ["mainEntity", "hasTitle"], after, changeNotes)
+                break
+            case "https://id.kb.se/changenote/primarypublication":
+            case "https://id.kb.se/changenote/serialtermination":
+                historicEmbellish(instanceAfterChange, ["publication"], after, changeNotes)
+                break
+            case "https://id.kb.se/changenote/intendedaudience":
+                historicEmbellish(instanceAfterChange, ["mainEntity", "intendedAudience"], after, changeNotes)
+                break
+            case "https://id.kb.se/changenote/ddcclassification":
+            case "https://id.kb.se/changenote/sabclassification":
+                historicEmbellish(instanceAfterChange, ["mainEntity", "classification"], after, changeNotes)
+                break
+            case "https://id.kb.se/changenote/serialrelation":
+                historicEmbellish(instanceAfterChange, ["mainEntity", "precededBy", "succeededBy"], after, changeNotes)
+                break
+            case "https://id.kb.se/changenote/primarycontribution": {
+                historicEmbellish(instanceAfterChange, ["mainEntity", "instanceOf", "contribution"], after, changeNotes)
+
```

Comment:
Lots of hardcoded domain terms here. Cf. https://github.com/libris/definitions/blob/develop/source/changenotes.ttl and https://github.com/libris/lxlviewer/blob/develop/vue-client/src/utils/changecategories.js for "less obscure" hardcoding? (Normally when data is repeated more than once in code, it's time to actually move it to the data.)